rfcs/0030-ensure-index-less-than-length-in-since/0030-ensure-index-less-than-length-in-since.md

```diff
@@ -1 +1,87 @@
+Status: Draft
+Author: Ian Yang
+Organization: Nervos Foundation
+Created: 2021-02-04
+---
+
+# Ensure That Index Is Less Than Length In the Input Since Field Using Epoch With Fraction
+
+## Abstract
+
+This document proposes to fix a bug in the CKB reference implementation. When a transaction input uses the epoch with fraction in the `since` field, the `value` must ensure that the index is less than the length.
+
+## Motivation
+
+The `value` part is an encoded number `epoch + index / length`. In CKB, every block is also assigned an epoch number with a fraction. If the block is the `I`th block in the epoch which epoch number is `E` and length is `L`, its epoch number with the fraction is `E + I / L` and `I` must be less than `L`. 
+
+To verify that a transaction is mature since an absolute epoch number with a fraction, the current implementation assumes that `index` must be less than `length` and checks that:
+
+```
+E > epoch
+or (E == epoch and I / L >= index/length)
+```
+
+This is not consistent when using relative flag in the `since` field, where the `value` is considered as a rational number and added to the cell start epoch.
+
+This document suggests adding a new rule to verify that when `since` uses epoch as the unit, it must ensure that index is less than the length.
+
+See RFC17 [Transaction valid since](../0017-tx-valid-since/0017-tx-valid-since.md) for details of the `since` field encoding patterns.
+
+## Specification
+
+### Since Absolute Epoch
+
+When an input `since` field is present, and
+
+* The `metric_flag` is epoch (01).
+* The `relative_flag` is absolute (0).
+
+Providing that `value` is the encoded number `epoch + index / length`, the transaction is valid only if `index` is less than `length` or both `index` and `length` are zero \*.
+
+> \* When both `index` and `length` are zero, `value` is converted to `epoch + 0 / 1`. 
+
+If the current epoch number with the fraction is `E + I / L`, the transaction is considered mature when
+
+```
+E + I / L >= epoch + index / length
+```
+
+using rational number operations.
+
+### Since Relative Epoch
+
+When an input `since` field is present, and
+
+* The `metric_flag` is epoch (01).
+* The `relative_flag` is relative (1).
+
+Providing that `value` is the encoded number `epoch + index / length`, the transaction is valid only if `index` is less than `length` or both `index` and `length` are zero \*.
+
+> \* When both `index` and `length` are zero, `value` is converted to `epoch + 0 / 1`. 
+
+If the cell creation epoch number with fraction is `e + i / l`, and the current epoch number with the fraction is `E + I / L`, the transaction is considered mature when
+
+```
+E + I / L >= e + i / l + epoch + index / length
+```
+
+using rational number operations.
+
+## Test Vectors
+
+When `since` uses the absolute epoch `99 + 360 / 180`, and the current epoch is `100 + 0 / 180`, the transaction is mature using the old consensus rule but is immature using the new rule.
```

Comment:
```suggestion
When `since` uses the absolute epoch `99 + 360 / 180`, and the current epoch is `100 + 0 / 180`, the transaction is mature using the old consensus rule but is invalid using the new rule.
```